test AnalysisFromFunction args order

HTian1997 opened the pull request:
Fixes #2503

Changes made in this Pull Request:
 - Correct argument order inspired by @lilyminium 
 - More compact coding for AnalysisFromFunction function
 - Add new testing on AnalysisFromFunction. 


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

HTian1997 commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -231,18 +231,16 @@ def __init__(self, function, trajectory=None, *args, **kwargs):
         if trajectory is None:
-            for arg in args:
+            # all possible places to find trajectory
+            possible = [arg for arg in args] + [arg for arg in six.itervalues(kwargs)]
```
I think the test function should be written in analysis/test_base.py right?

lilyminium replied:
Sorry, yes. And test `AnalysisFromFunction.args`. I was very distracted 🤦‍♀️

HTian1997 replied:
Don't fully understand "test `AnalysisFromFunction.args`". I have updated `test_base.py`. Please check whether that is what you mean.

HTian1997 commented:
Sorry I was trying to solve issue [#2470](https://github.com/MDAnalysis/mdanalysis/issues/2470) but I didn't know that I should PR in a different branch. Is that working on multiple PRs at the same time should be in different branches? 😕

RMeli commented:
@HTian1997 Yes, if you are working on different PRs, they should be in different branches, otherwise all the changes will be updated in the same PR.

orbeckst commented:
Different PRs please. The smaller the PR, the easier it is to review.

HTian1997 commented:
@orbeckst I have reverted the change. Please review.

lilyminium commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -162,6 +162,22 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+    assert_equal(isinstance(ans1.args[1], np.ndarray), True)
+    assert_equal(ans1.args[0], ans2.args[0])
+    assert_equal(ans1.args[1], ans2.args[1])
+    assert_equal(ans1.run().results, ans2.run().results)
```
Here you could have used ``assert isinstance(ans1.args[0], mda.core.groups.AtomGroup)`` instead.

In general I think the test should be a bit stronger, and test that `args[0] is protein` and same for the masses. It should also check that there is no mysterious `args[2]`. It would also be best to test that `ans2._trajectory is protein.universe.trajectory`, since that is what the code is actually looking for.

After that you probably don't need to actually `run()` the analysis.

lilyminium commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -162,6 +162,22 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+        return atomgroup.positions * masses
+    
+    protein = u.select_atoms('protein')
+    ans1 = base.AnalysisFromFunction(mass_xyz, u.trajectory, protein, protein.masses.reshape(-1, 1))
```
In my opinion, you don't need to have `ans1` here, since this test is for when the trajectory is not given. However, the tests don't actually have a case for when the trajectory is given, so it would be great if you made this a separate test and checked the argument order, and `ans1._trajectory` identity, again.

HTian1997 replied:
what do you mean by `made this a separate test and checked the argument order`, one function for testing case and another for testing argument order?

lilyminium replied:
As @orbeckst elaborated on below, yes, checking the arguments when a trajectory is not given should be done in a separate test to checking the arguments when a trajectory is given.

lilyminium commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -162,6 +162,22 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+    
+    protein = u.select_atoms('protein')
+    ans1 = base.AnalysisFromFunction(mass_xyz, u.trajectory, protein, protein.masses.reshape(-1, 1))
+    ans2 = base.AnalysisFromFunction(mass_xyz, protein, protein.masses.reshape(-1, 1))
```
Please follow [PEP8](https://www.mdanalysis.org/UserGuide/contributing_code.html#code-formatting-in-python) line lengths when possible.

HTian1997 commented:
@lilyminium I have added testing on 1) multiple traj inputs; 2) check `args[0] is protein` and `args[1] is masses`; 3) no mysterious `args[2]`; 4) `ans2._trajectory is protein.universe.trajectory`; 5) deleted unnecessary `ans1`. I forgot to pay attention to the line length. Will change that in the next commit.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -162,6 +162,25 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
         assert_equal(results, ana.results)
 
 
+def test_AnalysisFromFunction_args_order(u):
```
As a general rule, one test function for one assertion _especially_ if you are testing different things such as different arguments.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -162,6 +162,25 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+    
+    protein = u.select_atoms('protein')
+    sub = u.atoms[:100]
+    #ans1 = base.AnalysisFromFunction(mass_xyz, u.trajectory, protein, protein.masses.reshape(-1, 1))
```
This should be one test function.

Each test should only create one `ans`.

orbeckst replied:
You might be able to write the three tests as a parametrized test with `@pytest.mark.paramterize` – however, if you haven't done this before, just write three separate tests to see if there's enough common structure and then we can help with condensing in a parametrized test.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -162,6 +162,25 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+    sub = u.atoms[:100]
+    #ans1 = base.AnalysisFromFunction(mass_xyz, u.trajectory, protein, protein.masses.reshape(-1, 1))
+    # lack traj but can find in protein
+    ans2 = base.AnalysisFromFunction(mass_xyz, protein, protein.masses.reshape(-1, 1))
```
This can be the center of the second test.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -162,6 +162,25 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+    # lack traj but can find in protein
+    ans2 = base.AnalysisFromFunction(mass_xyz, protein, protein.masses.reshape(-1, 1))
+    # multiple traj input
+    ans3 = base.AnalysisFromFunction(mass_xyz, protein, sub, protein.masses.reshape(-1, 1))
```
Third test.

lilyminium replied:
Also, just a note, `sub.universe.trajectory is protein.universe.trajectory` i.e. the [trajectories are the same object in memory](http://net-informations.com/python/iq/is.htm), so your test for multiple trajectory inputs would always succeed but for the wrong reasons. To test that you are always finding the *first* trajectory, you would need to create different universes and check `ans._trajectory is protein.universe.trajectory`.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -162,6 +162,25 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+    assert (ans2.args[0] == protein) and not (ans2.args[1] - protein.masses.reshape(-1, 1)).any()
+    assert ans2._trajectory == protein.universe.trajectory
+    assert ans3.args[0] == protein and ans3.args[1] == sub
+   
```
Inspecting `ans` is good. (And  multiple `assert` for the _same_ `ans` are fine here.)

I would also add a short `ans.run(stop=5)` to make sure it really runs for a few steps. You should get something in `ans.result`. Just do a mean or a sum over whatever values you find there and check that value as an additional test.

HTian1997 commented:
@orbeckst Separated into three functions. 0) self-defined function; 1) test args length and check `ans.run()` result; 2) check args content; 3) check whether it can catch first traj given multiple trajs.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -161,6 +161,31 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+    protein = u.select_atoms('protein')
+    masses = protein.masses.reshape(-1, 1)
+    ans = base.AnalysisFromFunction(mass_xyz, protein, masses)
+    assert (ans.args[0] == protein) and not (ans.args[1] - masses).any()
```
`(ans.args[1] - masses).any()` tries to compare floating point numbers to 0. That's fragile. Avoid doing this.

Furthermore, what is going to happen if the code is still wrong? Are you able to calculate `ans.args[1] - masses` or will this throw a TypeError or something like that? In this case, this is not a clean failure. Please explain a bit more what the reasoning is behind this test.

HTian1997 replied:
if the code is wrong, that is, `ans.args[0] is masses` and `ans.args[1] is protein`, will return `AssertionError`. Do we need to deal with this error, like `pytest.raises` ?

orbeckst replied:
`assert (ans.args[0] is protein) and (ans.args[1] is masses)` is better as you'll always get True/False answer and not an exception.

I'll call it a  "clean failure" when the `assert` raises `AssertionError` because of whatever you're asserting. Any other exception that gets raised will also make the test fail but in a way that we did not anticipate or explicitly test and it will show up differently when pytest runs. It's a sign of a bug, not a controlled test.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -161,6 +161,31 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
+    masses = protein.masses.reshape(-1, 1)
+    another = mda.Universe(TPR, XTC).select_atoms("protein")
+    ans = base.AnalysisFromFunction(mass_xyz, another, protein, masses)
+    assert ans._trajectory == another.universe.trajectory
```
@lilyminium is this what it should be testing here?

lilyminium replied:
Yes, although if you use [is instead of ==](http://net-informations.com/python/iq/is.htm) you could avoid importing new files. In general, using `is` would simplify these tests.

Also, `mass_xyz` wouldn't actually run here.

HTian1997 replied:
okay, I will change `==` to `is`. Even `mass_xyz` couldn't run here, we can still test the args order, is this a problem?

HTian1997 replied:
@lilyminium I have revised `mass_xyz` which can run given two trajs.

HTian1997 commented:
@orbeckst  I combined previous three functions into one as there is only one `ans` and only testing on `ans`.

HTian1997 commented:
I really need to learn git command to amend instead of commit again and again on GitHub website 😭. So sorry about that.

RMeli commented:
@HTian1997 As long as your commit messages are sensible and informative, I think it's better to keep the whole history of your work. Rewriting/altering git history is a dangerous game, in my experience... They will be squashed together when merging, so they will not all end up in the MDAnalysis history anyway! ;)

HTian1997 commented:
> @HTian1997 As long as your commit messages are sensible and informative, I think it's better to keep the whole history of your work. Rewriting/altering git history is a dangerous game, in my experience... They will be squashed together when merging, so they will not all end up in the MDAnalysis history anyway! ;)

HTian1997 commented:
accidentally clicked close...

lilyminium commented:
Thanks @HTian1997! Please [update CHANGELOG](https://www.mdanalysis.org/UserGuide/contributing_code.html#documenting-your-code) as well with your username and the issue that you fixed :-)

lilyminium commented on package/CHANGELOG:
```diff
@@ -134,6 +135,8 @@ Changes
   * analysis.hbonds.HydrogenBondAnalysis is deprecated in 1.0 (remove in 2.0)
 
+Testsuite
+  * add test for args order of base.AnalysisFromFunction (PR #2594)
```
Tests don't typically get their own entry in the CHANGELOG. Otherwise, looks good to me!

orbeckst commented:
Thank you @HTian1997 ! We're waiting for CI to come back green and then we squash-merge.

HTian1997 commented:
@lilyminium @orbeckst Thank you for the help!

orbeckst commented:
@HTian1997 congratulations to your first merged PR! Thank you. 🥇
